ambry-network/src/main/java/com/github/ambry/network/http2/AmbrySendToHttp2Adaptor.java

```diff
@@ -71,9 +74,24 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
     }
     DefaultHttp2HeadersFrame headersFrame = new DefaultHttp2HeadersFrame(http2Headers, false);
     ctx.write(headersFrame);
-    ByteBuf dataContent = send.content();
-    DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(dataContent, true);
-    // Caller should call writeAndFlush().
-    ctx.write(dataFrame, promise);
+
+    // Referencing counting for derived {@link ByteBuf}: https://netty.io/wiki/reference-counted-objects.html#derived-buffers
+    try {
+      while (send.content().readableBytes() > maxFrameSize) {
+        ByteBuf slice = send.content().readSlice(maxFrameSize);
+        slice.retain();
+        DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(slice, false);
+        ctx.write(dataFrame, promise);
+      }
+      // The last slice
+      ByteBuf slice = send.content().readSlice(send.content().readableBytes());
+      slice.retain();
+      DefaultHttp2DataFrame dataFrame = new DefaultHttp2DataFrame(slice, true);
+      ctx.writeAndFlush(dataFrame, promise);
+    } catch (Exception e) {
+      logger.error("Error while processing frames. Channel: {}", ctx.channel(), e);
+    } finally {
+      send.content().release();
```

Comment:
And here since send.content() is never passed to ctx.write method, send.content()'s refCnt won't be decreased, is that the reason why you need to call release here?

Reply:
That's correct. See examples from netty doc:  https://netty.io/wiki/reference-counted-objects.html#derived-buffers